File: products/windshift/internal/routes/integrations.go
package routes

import (
	"net/http"

	"windshift/internal/handlers"
)

// IntegrationHandlers groups integration-related handlers.
type IntegrationHandlers struct {
	Provider    *handlers.IntegrationProviderHandler
	OAuth       *handlers.IntegrationOAuthHandler
	ItemLinks   *handlers.IntegrationItemLinksHandler
	TodoistSync *handlers.TodoistSyncHandler
}

// RegisterIntegrationRoutes registers integration provider, OAuth, and item link routes.
func RegisterIntegrationRoutes(deps *Deps) {
	api := deps.API
	auth := deps.AuthMiddleware.RequireAuth
	admin := deps.PermissionMiddleware.RequireSystemAdmin()

	// Admin provider management
	api.HandleH("GET /admin/integration-providers", admin(http.HandlerFunc(deps.Integrations.Provider.GetProviders)))
	api.HandleH("POST /admin/integration-providers", admin(http.HandlerFunc(deps.Integrations.Provider.CreateProvider)))
	api.HandleH("GET /admin/integration-providers/{id}", admin(http.HandlerFunc(deps.Integrations.Provider.GetProvider)))
	api.HandleH("PUT /admin/integration-providers/{id}", admin(http.HandlerFunc(deps.Integrations.Provider.UpdateProvider)))
	api.HandleH("DELETE /admin/integration-providers/{id}", admin(http.HandlerFunc(deps.Integrations.Provider.DeleteProvider)))

	// OAuth flow
	api.HandleH("GET /integrations/oauth/{slug}/start", auth(http.HandlerFunc(deps.Integrations.OAuth.StartOAuth)))
	api.Handle("GET /integrations/oauth/{slug}/callback", http.HandlerFunc(deps.Integrations.OAuth.OAuthCallback))

	// User connections
	api.HandleH("GET /users/me/integration-connections", auth(http.HandlerFunc(deps.Integrations.OAuth.GetUserConnections)))
	api.HandleH("GET /users/me/integration-connections/available", auth(http.HandlerFunc(deps.Integrations.OAuth.GetAvailableProviders)))
	api.HandleH("DELETE /users/me/integration-connections/{provider_id}", auth(http.HandlerFunc(deps.Integrations.OAuth.DisconnectProvider)))

	// Todoist personal-task sync
	api.HandleH("GET /users/me/todoist-sync", auth(http.HandlerFunc(deps.Integrations.TodoistSync.GetSync)))
	api.HandleH("PUT /users/me/todoist-sync", auth(http.HandlerFunc(deps.Integrations.TodoistSync.UpdateSync)))
	api.HandleH("GET /users/me/todoist-sync/projects", auth(http.HandlerFunc(deps.Integrations.TodoistSync.GetProjects)))
	api.HandleH("POST /users/me/todoist-sync/run", auth(http.HandlerFunc(deps.Integrations.TodoistSync.RunSync)))

	// Item links
	api.HandleH("GET /items/{id}/integration-links", auth(http.HandlerFunc(deps.Integrations.ItemLinks.GetItemLinks)))
	api.HandleH("POST /items/{id}/integration-links", auth(http.HandlerFunc(deps.Integrations.ItemLinks.CreateItemLink)))
	api.HandleH("DELETE /item-integration-links/{linkId}", auth(http.HandlerFunc(deps.Integrations.ItemLinks.DeleteItemLink)))
	api.HandleH("POST /item-integration-links/{linkId}/refresh", auth(http.HandlerFunc(deps.Integrations.ItemLinks.RefreshItemLink)))
	api.HandleH("GET /items/{id}/integration-search", auth(http.HandlerFunc(deps.Integrations.ItemLinks.SearchPages)))
}

File: products/windshift/frontend/src/lib/api/integrations.js
import { fetchAPI } from './core.js';
import { createCrudClient } from './createCrudClient.js';

// Integration providers (admin management)
export const integrationProviders = createCrudClient('/admin/integration-providers');

// User integration connections
export const userIntegrations = {
  getConnections: (options) => fetchAPI('/users/me/integration-connections', options),
  getAvailableProviders: (options) =>
    fetchAPI('/users/me/integration-connections/available', options),
  disconnect: (providerId) =>
    fetchAPI(`/users/me/integration-connections/${providerId}`, {
      method: 'DELETE',
    }),
  startOAuth: (slug) => fetchAPI(`/integrations/oauth/${slug}/start`),
};

// Todoist personal-task sync
export const todoistSync = {
  get: () => fetchAPI('/users/me/todoist-sync'),
  update: (data) =>
    fetchAPI('/users/me/todoist-sync', {
      method: 'PUT',
      body: JSON.stringify(data),
    }),
  getProjects: () => fetchAPI('/users/me/todoist-sync/projects'),
  run: () => fetchAPI('/users/me/todoist-sync/run', { method: 'POST' }),
};

// Item integration links
export const itemIntegrationLinks = {
  get: (itemId, options) => fetchAPI(`/items/${itemId}/integration-links`, options),
  create: (itemId, data) =>
    fetchAPI(`/items/${itemId}/integration-links`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  delete: (linkId) =>
    fetchAPI(`/item-integration-links/${linkId}`, {
      method: 'DELETE',
    }),
  refresh: (linkId) =>
    fetchAPI(`/item-integration-links/${linkId}/refresh`, {
      method: 'POST',
    }),
  search: (itemId, query, providerId) =>
    fetchAPI(
      `/items/${itemId}/integration-search?q=${encodeURIComponent(query)}&provider_id=${providerId}`
    ),
};
